src/ui/public/state_management/state_storage/hashed_item_store.js

```diff
@@ -0,0 +1,165 @@
+ *
+ * Consideration 2: We can't tell when we've hit the browser history limit
+ *
+ * Because some of our persisted history states may no longer be referenced by the browser history,
+ * and we have no way of knowing which ones, we have no way of knowing whether we've persisted a
+ * number of accessible states beyond the browser history length limit.
+ *
+ * More fundamentally, the browser history length limit is a browser implementation detail, so it
+ * can change from browser to browser, or over time. Respecting this limit would introduce a lot of
+ * (unnecessary?) complexity.
+ *
+ * For these reasons, HashedItemStore doesn't concern itself with this constraint.
+ */
+
+export default class HashedItemStore {
+
+  /**
+   * HashedItemStore uses objects called indexed items to refer to items that have been persisted
+   * in sessionStorage. An indexed item is shaped {hash, touched}. The touched date is when the item
+   * was last referenced by the browser history.
+   */
+  constructor(sessionStorage = window.sessionStorage) {
+    this._sessionStorage = sessionStorage;
+
+    // Store indexed items in descending order by touched (oldest first, newest last). We'll use
+    // this to remove older items when we run out of storage space.
+    this._items = [];
+    // Associate item hashes with the corresponding indexed items. We'll use this to quickly look
+    // up an item and update its touched date when it reoccurs in the browser history.
+    this._itemIndex = {};
+
+    // Build index from the persisted index. This happens when we re-open a closed tab.
+    const persistedItemIndex = this._sessionStorage.getItem(HashedItemStore.PERSISTED_INDEX_KEY);
+
+    if (persistedItemIndex) {
+      this._itemIndex = JSON.parse(persistedItemIndex) || {};
+      Object.keys(this._itemIndex).forEach(hash => {
+        const indexedItem = this._itemIndex[hash];
+        this._items.push(indexedItem);
+      });
+
+      // Order items by touched date (oldest first, newest last).
+      this._items.sort((a, b) => {
+        return a.touched - b.touched;
+      });
+    }
+  }
+
+  setItem(hash, item) {
+    const isItemPersisted = this._persistItem(hash, item);
+
+    if (isItemPersisted) {
+      this._touchHash(hash);
+    }
+
+    return isItemPersisted;
+  }
+
+  getItem(hash) {
+    const item = this._sessionStorage.getItem(hash);
+    this._touchHash(hash);
+    return item;
+  }
+
+  _persistItem(hash, item) {
+    try {
+      this._sessionStorage.setItem(hash, item);
+      return true;
+    } catch (e) {
+      // If there was an error then we need to make some space for the item.
+      if (this._items.length === 0) {
+        // If there's nothing left to remove, then we've run out of space and we're trying to
+        // persist too large an item.
+        return false;
+      }
+
+      // We need to try to make some space for the item.
+
+      // Remove oldest item from index.
+      const indexedItem = this._items.shift();
```

Comment:
I think this would be great a great chunk to pull out into `this._removeItem()` or something. This section might read a little smoother then.